README.md
# IAPChecker

An iOS in-application purchase receipt verifier.

## Contributors

* [objectiveSee](https://github.com/objectiveSee)


## How To

### Install

```
npm install iapc
```

### Use

```
var IAPChecker = require('iapc');

/*
	Create an instance of IAPChecker with a password used when verifying an 
	auto renewable receipt and a boolean to verify the receipt at Apple
	production or sandbox server.
*/

var iapc = new IAPChecker('MyPassword'[, true]);

checkReceipt(
	MyReceipt,
	function(err, data) {
		if (err) return console.log(err);
		
		// receipt is valid
	}
);
```

## API

* **`IAPChecker(password, production)`**

	Create an instance of IAPChecker. All parameters are optional. `password` must be set if you want to verify an auto-renewable subscription receipt, and set `production` to work in production environment.

* **`checkAutoRenewReceipt(receipt, callback)`**

	Check the auto renewable receipt. A password must be set when using this method. A response will be send back to the callback.

* **`checkReceipt(receipt, callback)`**

	Check receipt. A response will be send back to the callback.

* **`setPassword(password)`**

	Set the instance password.

CHANGE_LOG.md
# Change log

## v0.2.0

* update `request` module version (2.21.0 to 2.27.0)
* code refactoring
* add [objectiveSee](https://github.com/objectiveSee) pull request

## v0.1.1

* update package module version(s)
* add **david-dm.org** support

## v0.1.0

* first release

lib/checker.js
/**
 * Author: Ismael Gorissen
 * Date: 25/04/13 11:17
 * Company: PinchProject
 */

var debug = require('debug')('inapp'),
    request = require('request');

var _hosts = {
        production: 'buy.itunes.apple.com',
        sandbox: 'sandbox.itunes.apple.com'
    },
    _responseCodes = {
        unknown: {
            message: 'Unknown status code',
            valid: false,
            error: true
        },
        0: {
            message: "Active",
            valid: true,
            error: false
        },
        21000: {
            message: "App store could not read",
            valid: false,
            error: true
        },
        21002: {
            message: "Data was malformed",
            valid: false,
            error: true
        },
        21003: {
            message: "Receipt not authenticated",
            valid: false,
            error: true
        },
        21004: {
            message: "Shared secret does not match",
            valid: false,
            error: true
        },
        21005: {
            message: "Receipt server unavailable",
            valid: false,
            error: true
        },
        21006: {
            message: "Receipt valid but sub expired",
            valid: false,
            error: false
        },
        21007: {
            message: "Sandbox receipt sent to Production environment",
            valid: false,
            error: true,
            redirect: true
        },
        21008: {
            message: "Production receipt sent to Sandbox environment",
            valid: false,
            error: true
        }
    };

function IAPChecker(secret, production) {
    this.secret = secret;
    this.production = production ? production : false;
    this.host = this.production ? _hosts.production : _hosts.sandbox;
}

IAPChecker.prototype = {
    checkAutoRenewReceipt: checkAutoRenewReceipt,
    checkReceipt: checkReceipt,
    setSecret: setSecret
};

function checkAutoRenewReceipt(receipt, done) {
    var data = {
        'receipt-data': '',
        password: this.secret
    };

    checkWithRetry(this, data, receipt, done);
}

function checkReceipt(receipt, done) {
    var data = {
        'receipt-data': ''
    };

    checkWithRetry(this, data, receipt, done);
}

function checkWithRetry(self, receiptData, receipt, done) {
    check(self, receiptData, receipt, function (err, valid, message, data) {
        if (!err) {
            if ((data.status == 21007) && (self.host === _hosts.production)) {
                self.host = _hosts.sandbox;

                check(self, receiptData, receipt, function (err, valid, message, data) {
                    done(err, valid, message, data);
                });
            } else {
                done(err, valid, message, data);
            }
        } else {
            done(err);
        }
    });
}

function check(self, data, receipt, done) {
    var buffer,
        encoded,
        postData;

    buffer = new Buffer(receipt);
    encoded = buffer.toString('base64');

    data['receipt-data'] = encoded;

    postData = JSON.stringify(data);

    var uri = 'https://' + self.host + '/verifyReceipt';

    request({
        uri: uri,
        method: 'POST',
        body: postData,
        headers: {
            'content-type': 'application/x-www-form-urlencoded',
            'content-length': postData.length
        },
        timeout: 10000
    }, function (err, res, body) {
        if (!err) {
            try {
                body = JSON.parse(body);

                if (res.statusCode == 200) {
                    processStatus(body, done);
                } else {
                    return done(null, false, body);
                }
            } catch (ex) {
                done(ex);
            }
        } else {
            done(err);
        }
    });
}

function processStatus(data, done) {
    var response;

    if (_responseCodes.hasOwnProperty(data.status)) {
        response = _responseCodes[data.status];
    } else {
        response = _responseCodes.unknown
    }

    done(null, response.valid, response.message, data);
}

function setSecret(secret) {
    this.secret = secret;
}

module.exports = IAPChecker;

lib/iapc.js
/**
 * Author: Ismael Gorissen
 * Company: PinchProject
 */

var request = require('request'),
    debug = require('debug')('IAPC');
var Configurator = require('./modules/configurator');

var HOSTS = Configurator.load('hosts');
var RESPONSE_CODES_MESSAGE = Configurator.load('response_codes_message');

/**
 *
 * @param password
 * @param production
 * @constructor
 */
function IAPChecker(password, production) {
    this.password = password;
    this.production = production ? true : false;
    this.host = this.production ? HOSTS.production : HOSTS.sandbox;
}

/**
 *
 * @param password
 */
IAPChecker.prototype.setPassword = function (password) {
    this.password = password;
};

/**
 *
 * @param receipt
 * @param callback
 */
IAPChecker.prototype.checkAutoRenewReceipt = function (receipt, callback) {
    var self = this;

    var data = {
        'receipt-data': self._encodeReceiptInBase64(receipt),
        password: self.password
    };

    self._check(data, callback);
};

/**
 *
 * @param receipt
 * @param callback
 */
IAPChecker.prototype.checkReceipt = function (receipt, callback) {
    var self = this;

    var data = {
        'receipt-data': self._encodeReceiptInBase64(receipt)
    };

    self._check(data, callback);
};

/**
 *
 * @param receiptData
 * @param callback
 * @private
 */
IAPChecker.prototype._check = function (receiptData, callback) {
    var self = this;

    var body = JSON.stringify(receiptData);
    var options = {
        uri: self._hostUri(),
        method: 'POST',
        body: body,
        headers: {
            'content-type': 'application/x-www-form-urlencoded',
            'content-length': body.length
        },
        timeout: 10000
    };

    self._post(
        options,
        function (err, data) {
            if (err) return callback(err);

            if (data.status == 0)
                return callback(null, data);
            if (data.status != 21007 && data.status != 21008)
                return callback(new Error(RESPONSE_CODES_MESSAGE['' + data.status + '']));

            if (data.status == 21007)
                self.host = HOSTS.sandbox;
            if (data.status == 21008)
                self.host = HOSTS.production;

            self._check(receiptData, callback);
        }
    );
};

/**
 *
 * @param options
 * @param callback
 * @private
 */
IAPChecker.prototype._post = function (options, callback) {
    request(
        options,
        function (err, response, body) {
            if (err) return callback(err);

            try {
                body = JSON.parse(body);

                if (response.statusCode == 200) {
                    callback(null, body);
                } else {
                    err = new Error('Expected HTTP status code "200", got "' + response.statusCode + '"');
                    err.body = body;
                    callback(err);
                }
            } catch (ex) {
                return callback(new Error('Response body can not be parsed'));
            }
        }
    )
};

/**
 *
 * @returns {string}
 * @private
 */
IAPChecker.prototype._hostUri = function () {
    return 'https://' + this.host + '/verifyReceipt';
};

/**
 *
 * @param receipt
 * @returns {*|String}
 * @private
 */
IAPChecker.prototype._encodeReceiptInBase64 = function (receipt) {
    return new Buffer(receipt).toString('base64');
};

module.exports = IAPChecker;